tendrl/monitoring_integration/sync/dashbaord_sync.py

```diff
@@ -47,120 +59,135 @@ def refresh_dashboard(self):
-                    cluster_details["sds_name"]
-                )
+    def create_dashboard(self, resource_type, resources):
+        resource_json = {}
+        # create dashboard
+        if resources:
+            # Create operation takes more excution time than update
+            # Create dashboard for one resource using that do update for
+            # other resources
+            resource = resources.pop()
+            resource_json = alert_dashboard.create_resource_dashboard(
+                resource_type,
+                resource
+            )
+        # After creating first panel update the other panels
+        if resources and resource_json:
+            # update dashboard
+            rows = resource_json["dashboard"]["rows"]
+            most_recent_panel_id = rows[-1]["panels"][-1]["id"]
+            resource_json = alert_dashboard.add_panel(
+                resources,
+                resource_type,
+                resource_json,
+                most_recent_panel_id
+            )
+        return resource_json
 
-    def create_gluster_resource(self, resources, resource_type,
-                                integration_id, sds_name):
-        for resource in resources:
-            try:
-                if resource_type == "volumes":
-                    resource_name = str(resource["name"])
-                elif resource_type == "hosts":
-                    resource_name = str(resource["fqdn"]).replace(".", "_")
-                elif resource_type == "bricks":
-                    resource_name = "%s|%s:%s" % (
-                        str(resource["vol_name"]),
-                        resource["hostname"],
-                        resource["brick_path"].replace("|", "/")
-                    )
-            except KeyError:
-                logger.log("debug", NS.get("publisher_id", None),
-                           {'message': "Failed to get resource {} "
-                            "details".format(resource)})
-                continue
-            response = alert_dashboard.add_panel(
-                integration_id,
+    def update_dashboard(self, resource_type, resources, resource_json):
+        no_changes_in_dashboard = False
+        new_resources, alert_dashboard_json, most_recent_panel_id = \
+            alert_dashboard.check_duplicate(
+                copy.deepcopy(resource_json),
+                resources,
+                resource_type
+            )
+        if new_resources and len(
+            alert_dashboard_json["dashboard"]["rows"]
+        ) > 0:
+            alert_dashboard_json = alert_dashboard.add_panel(
+                new_resources,
                 resource_type,
-                resource_name,
-                sds_name
+                alert_dashboard_json,
+                most_recent_panel_id
             )
-            if response:
-                self.log_message(response, resource_name, resource_type)
+        elif len(resource_json["dashboard"]["rows"]) == \
```

Comment:
So now we check the if the length of rows in resource_json same as length of alert dashboard rows, its like no changes. correct? This takes care of jombie rows as well in alert dashboard?

Reply:
When we call a check_duplicate function there all x-y comparision will happen so all zombie panels will be removed and passed to this place. Here any changes in newly formed JSON with old json then only we are posting it into grafana. No resource newly added and something deleted the old json is won't match with new one also if anything is newly added in that case also new json is different so i will allow updating.

Reply:
ok